### DISCUSSION TITLE
depends_on ordering seems wrong

### DISCUSSION THREAD

**Discussion by mjmeli** [containrrr/watchtower#1261] [Q&A] [answered]
I've got a compose running with two containers, one a VPN container (gluetun) and one a torrent container (qbittorrent). The torrent container routes through the VPN container for networking (`network_mode: "service:gluetun"`). When gluetun updates, I need qbittorrent to shutdown first, then gluetun shutdown, then run the update, then gluetun restarts, then qbittorrent restarts.

I am trying to use the [Linked Containers](https://containrrr.dev/watchtower/linked-containers/) functionality for this.

However, it looks like the dependency link will only work one way, and not the way I need it.

Here is a my condensed base docker-compose file

    services:
      gluetun:
        image: qmcgaw/gluetun:latest
        container_name: gluetun
        cap_add:
          - NET_ADMIN
        network_mode: bridge
        ports:
          ...
        volumes:
          ...
        environment:
          ...
        restart: unless-stopped

      qbittorrent:
        image: lscr.io/linuxserver/qbittorrent
        container_name: qbittorrent
        privileged: true
        environment:
          ...
        volumes:
          ...
        network_mode: "service:gluetun"
        restart: unless-stopped
        depends_on:
          - gluetun

It isn't really clear from the documentation which container gets the label, so I've tried adding the label to both containers, but neither works.

**Label on qbittorrent**

    services:
      gluetun:
        ...

      qbittorrent:
        ...
        labels:
          com.centurylinklabs.watchtower.depends-on: gluetun

Results in the following when gluetun updates.

    time="2022-03-29T04:00:28-04:00" level=info msg="Found new qmcgaw/gluetun:latest image (69ed4f2da6b5)"
    time="2022-03-29T04:01:48-04:00" level=info msg="Stopping /gluetun (1f18b5a2f252) with SIGTERM"
    time="2022-03-29T04:01:53-04:00" level=info msg="Stopping /qbittorrent (08ddd2361630) with SIGTERM"
    time="2022-03-29T04:02:02-04:00" level=info msg="Creating /qbittorrent"
    time="2022-03-29T04:02:03-04:00" level=error msg="Error response from daemon: No such container: 1f18b5a2f25268b6542e15241d20e44df4c126b8871af3c8fe5e982c743efe1e"
    time="2022-03-29T04:02:03-04:00" level=info msg="Creating /gluetun"

**Label on gluetun**

    services:
      gluetun:
        ...
        labels:
          com.centurylinklabs.watchtower.depends-on: qbittorrent

      qbittorrent:
        ...

Results in the following when gluetun updates

    time="2022-03-31T04:00:28-04:00" level=info msg="Found new qmcgaw/gluetun:latest image (7e7108368891)"
    time="2022-03-31T04:03:23-04:00" level=info msg="Stopping /gluetun (c04867adc356) with SIGTERM"
    time="2022-03-31T04:03:28-04:00" level=info msg="Creating /gluetun"

And results in the following when qbitorrent updates

    time="2022-03-25T04:00:35-04:00" level=info msg="Found new lscr.io/linuxserver/qbittorrent:latest image (edce4fb7a1b0)"
    time="2022-03-25T04:02:04-04:00" level=info msg="Stopping /gluetun (54b986556062) with SIGTERM"
    time="2022-03-25T04:02:09-04:00" level=info msg="Stopping /qbittorrent (d76eee15fcd2) with SIGTERM"
    time="2022-03-25T04:02:17-04:00" level=info msg="Creating /qbittorrent"
    time="2022-03-25T04:02:18-04:00" level=error msg="Error response from daemon: No such container: 54b986556062b61d5dbd02b8c486c491259d8a7df2f21655201905275624a0fc"
    time="2022-03-25T04:02:18-04:00" level=info msg="Creating /gluetun"

**The Issue**

So the label on either container isn't achieving the result I need.

If I think through each dependency link and update case, I'd get the following:

| Dependency Link | Expected sequence when qbittorrent updates | Expected sequence when gluetun updates |
| ------------------- | -------------------------------------------------- | ----------------------------------------------- |
| Q depends on G | stop Q, start Q | stop Q, stop G, start G, start Q |
| G depends on Q | stop G, stop Q, start Q, start G | stop G, start G |

Based on the results above, this isn't being followed. The "G depends on Q" behavior is happening when the label is on gluetun, but the "Q depends on G" behavior is **backwards** when the label is on qbittorrent. The fact that in both labeling cases, gluetun is being stopped before qbittorrent indicates one of them is wrong.

I think this is a bug but I was wondering if someone could check this before I open an issue. What container exactly should I be putting the label on?

**Comment by piksel** [comment, marked as answer]
First of all, are you running `v1.4.0` of watchtower? There *were* a bug  that reversed the restart order in earlier versions of watchtower.

You should be putting the label on Q (depends on G). That would mean that "Q needs G to run, so stop it before G and start it after G".

**Reply by mjmeli** [reply]
Yes, I'm using v1.4.0. 

I had assumed that the label goes on Q, which is why I started there, but that wasn't working which is why I experimented the other way.

**Reply by mjmeli** [reply]
Well, I've done some additional testing where I manually update the gluetun VPN container, and simple stopping and then restarting the qBittorrent doesn't work, as restarting it does not update the container reference for the network. I have to re-create the container instead.

Perhaps watchtower is doing it in the correct order, but the logs are coming out of order due to some error starting containers.

Either way, my use-case doesn't seem supported so I have to go ahead with just disabling auto updates on the VPN container.
